### PR TITLE
Add ability to generate test users using Whistler API

### PR DESCRIPTION
**Pull request by taymoork2**
Fixes [SPARK-193921](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-194931)

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: packages/node_modules/@webex/test-users/src/whistler.js [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+      emailAddress: res.body.responseMetaData.name,
+      displayName: res.body.responseMetaData.webExUserName,
+      token: res.body.responseMetaData.ciAccessToken,
+      reservationUrl: res.body.reservationUrl
+    }, res.body.responseMetaData));
+}
+
+/**
+ *
+ * @param {Object} options
```
Comment (lalli-flores) [inline packages/node_modules/@webex/test-users/src/whistler.js:186]: Can you document expected options?

File: packages/node_modules/@webex/test-users/test/integration/spec/whistler.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    it('removes a test user if no access token is available', () => createWhistlerTestUser()
+      .then(prune)
+      .then(async (u) => {
+        const res = await removeWhistlerTestUser(u);
+
+        assert.equal(res.statusCode, 204);
+      }));
+
+    it('removes a test user with an access token set to undefined', () => createWhistlerTestUser()
```
Comment (lalli-flores) [inline packages/node_modules/@webex/test-users/test/integration/spec/whistler.js:58]: If there is no access token, how do you know what user to remove? or what gets removed?

File: packages/node_modules/@webex/test-users/src/whistler.js [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+  }
+
+  if (token.token_type && token.access_token) {
+    token.authorization = `${token.token_type} ${token.access_token}`;
+  }
+
+  return token;
+}
+
+let clientToken;
```
Comment (lalli-flores) [inline packages/node_modules/@webex/test-users/src/whistler.js:30]: Why would you want to cache the user? Isn't the expectation that I will get a new user every time I call "create user"?

File: packages/node_modules/@webex/test-users/src/whistler.js (hunk 1 above)
Comment (taymoork2) [reply packages/node_modules/@webex/test-users/src/whistler.js:186]: It's the entire user object so there's too much to document, also options isnt the best name, but i'm keeping it aligned with what's in `index.js`

File: packages/node_modules/@webex/test-users/src/whistler.js [hunk 4]
```diff
@@ -0,0 +1,159 @@
+import btoa from 'btoa';
+import {request} from '@webex/http-core';
+
+/**
+ * Fetches credentials to talk to the test_users_s endpoint
+ *
+ * Caches result in `clientToken` variable for multiple runs
```
Comment (lalli-flores) [inline packages/node_modules/@webex/test-users/src/whistler.js:7]: I think this is no longer in the code

File: packages/node_modules/@webex/test-users/src/whistler.js [hunk 5]
```diff
@@ -0,0 +1,159 @@
+import btoa from 'btoa';
+import {request} from '@webex/http-core';
+
+/**
+ * Fetches credentials to talk to the test_users_s endpoint
```
Comment (lalli-flores) [inline packages/node_modules/@webex/test-users/src/whistler.js:5]:
```suggestion
 * Fetches credentials to talk to the test users's endpoint
```

File: packages/node_modules/@webex/test-users/src/whistler.js [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+      // us to send the auth header *without including "Basic "* before the
+      // token string
+      // authorization: `Basic + ${btoa(`${clientId}:${clientSecret}`)}`
+      authorization: btoa(`${clientId}:${clientSecret}`)
+    }
+  }))
+  .then((res) => `${res.body.token_type} ${res.body.access_token}`);
+
+/**
+ * @typedef {Object} AccessTokenObject
```
Comment (lalli-flores) [inline packages/node_modules/@webex/test-users/src/whistler.js:57]: I dont think this object exists anymore

File: packages/node_modules/@webex/test-users/src/whistler.js [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+ * @property {Number} token.expires_in
+ * @property {string} token.token_type
+ * @property {string} token.refresh_token
+ * @property {Number} token.refresh_token_expires_in
+ * @property {string} token.expires
+ * @property {string} token.refresh_token_expires
+ */
+
+/**
+  * @typedef {Object} TestUserObject
```
Comment (lalli-flores) [inline packages/node_modules/@webex/test-users/src/whistler.js:68]: I think there are a lot more properties now, can you update this to match?

File: packages/node_modules/@webex/test-users/src/whistler.js (hunk 1 above)
Comment (lalli-flores) [reply packages/node_modules/@webex/test-users/src/whistler.js:148]: There is a type definition above for `TestUserObject`, if that is what it's supposed to be then, can you update the correct type?
Maybe rename the argument to `user`

File: packages/node_modules/@webex/test-users/test/integration/spec/whistler.js [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+assert.isTestUser = (user) => {
+  assert.isDefined(user, 'user is defined');
+  assert.isDefined(user.displayName, 'user.displayName is defined');
+  assert.isDefined(user.emailAddress, 'user.emailAddress is defined');
+  assert.isDefined(user.password, 'user.password is defined');
+  assert.isDefined(user.reservationUrl, 'user.reservationUrl is defined');
+};
+
+describe('test-users', () => {
```
Comment (lalli-flores) [inline packages/node_modules/@webex/test-users/test/integration/spec/whistler.js:22]: Can you add tests to show what happens when there's a failure?

File: packages/node_modules/@webex/test-users/src/whistler.js (hunk 5 above)
Comment (adamweeks) [reply packages/node_modules/@webex/test-users/src/whistler.js:5]: This whistler file doesn't talk to the `test_users_s` convo endpoint.

File: packages/node_modules/@webex/test-users/src/whistler.js [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+/**
+ * Creates a test user
+ * @param {CreateUserOptions} options
+ * @returns {Promise.<TestUserObject>}
+ */
+export default function createTestUser(options = {}) {
+  const clientId = options.clientId || process.env.WEBEX_CLIENT_ID;
+  const clientSecret = options.clientSecret || process.env.WEBEX_CLIENT_SECRET;
+  const machineAccount = options.machineAccount || process.env.WHISTLER_MACHINE_ACCOUNT;
```
Comment (adamweeks) [inline packages/node_modules/@webex/test-users/src/whistler.js:94]: This is not in the typedef

File: packages/node_modules/@webex/test-users/src/whistler.js [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+  const machinePassword = options.idbrokerUrl || process.env.WHISTLER_MACHINE_PASSWORD;
```
Comment (adamweeks) [inline packages/node_modules/@webex/test-users/src/whistler.js:95]:
```suggestion
  const machinePassword = options.machinePassword || process.env.WHISTLER_MACHINE_PASSWORD;
```

File: packages/node_modules/@webex/test-users/src/whistler.js [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+  * @property {string} emailAddress
+  * @property {string} displayName
+  */
+
+/**
+ * @typedef {Object} CreateUserOptions
+ * @param {Object} [options]
+ * @param {string} [options.clientId] defaults to WEBEX_CLIENT_ID
+ * @param {string} [options.clientSecret] defaults to WEBEX_CLIENT_SECRET
+ * @param {string} [options.conversationServiceUrl] defaults to WEBEX_TEST_USERS_CONVERSATION_SERVICE_URL
```
Comment (adamweeks) [inline packages/node_modules/@webex/test-users/src/whistler.js:79]: Don't need this

Comment (adamweeks) [reply packages/node_modules/@webex/test-users/src/whistler.js:79]: Perhaps it should be `WHISTLER_API_SERVICE_URL`

File: packages/node_modules/@webex/test-users/src/whistler.js [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+    clientId,
+    clientSecret,
+    machineAccount,
+    machinePassword,
+    idbrokerUrl,
+    orgId
+  })
+    .then((authorization) => request({
+      method: 'GET',
+      uri: `${process.env.WHISTLER_API_SERVICE_URL}/reservations/testUser?isAccessTokenRequired=true`,
```
Comment (adamweeks) [inline packages/node_modules/@webex/test-users/src/whistler.js:132]: Can we make this an option like we had with the `options.conversationServiceUrl`?

**Comment by taymoork2** [comment]
@adamweeks writing tests to check if `options.whistler` is working, is becoming more difficult than it should be.
Sinon is unable to stub/spy on methods within the `@webex/test-users` package that is a dependency in `@webex/test-helpers-test-users`, where the test is being created/added. It especially doesn't like trying to stub the exported whistler create user function.
Michael and I tried figuring out a solution, but were both stumped; and I showed Michael the integration tests for whistler were passing.
I don't think we should spend more time on this PR and get to work on getting the other repo setup

**Review by adamweeks — APPROVED** [review]

**Review by mwegman — APPROVED** [review]

**Pull request merged**